torchani/data/new.py

```diff
@@ -0,0 +1,458 @@
+        energies = torch.from_numpy(self.data_energies[index]).float()
+
+        return [species, coordinates, energies]
+
+    def __len__(self):
+        return self.length
+
+
+def collate_fn(data, chunk_threshold):
+    """Creates a batch of chunked data.
+    """
+
+    # unzip a batch of molecules (each molecule is a list)
+    batch_species, batch_coordinates, batch_energies = zip(*data)
+    batch_size = len(batch_species)
+
+    # padding - time: 13.2s
+    batch_species = torch.nn.utils.rnn.pad_sequence(batch_species,
+                                                    batch_first=True,
+                                                    padding_value=-1)
+    batch_coordinates = torch.nn.utils.rnn.pad_sequence(batch_coordinates,
+                                                        batch_first=True,
+                                                        padding_value=0)
+    batch_energies = torch.stack(batch_energies)
+
+    # sort - time: 0.7s
+    atoms = torch.sum(~(batch_species == -1), dim=-1, dtype=torch.int32)
+    sorted_atoms, sorted_atoms_idx = torch.sort(atoms, descending=False)
+
+    batch_species = torch.index_select(batch_species, dim=0, index=sorted_atoms_idx)
+    batch_coordinates = torch.index_select(batch_coordinates, dim=0, index=sorted_atoms_idx)
+    batch_energies = torch.index_select(batch_energies, dim=0, index=sorted_atoms_idx)
+
+    # get chunk size - time: 2.1s
+    output, count = torch.unique(atoms, sorted=True, return_counts=True)
+    counts = torch.cat((output.unsqueeze(-1).int(), count.unsqueeze(-1).int()), dim=-1)
+    chunk_size_list, chunk_max_list = split_to_chunks(counts, chunk_threshold=chunk_threshold * batch_size * 20)
+
+    # split into chunks - time: 0.3s
+    chunks_batch_species = torch.split(batch_species, chunk_size_list, dim=0)
+    chunks_batch_coordinates = torch.split(batch_coordinates, chunk_size_list, dim=0)
+
+    # truncate redundant padding - time: 1.3s
+    chunks_batch_species = trunc_pad(list(chunks_batch_species), padding_value=-1)
+    chunks_batch_coordinates = trunc_pad(list(chunks_batch_coordinates))
+
+    for i, c in enumerate(chunks_batch_coordinates):
+        chunks_batch_coordinates[i] = c.reshape(c.shape[0], -1, 3)
+
+    datas = [chunks_batch_species, chunks_batch_coordinates]
+
+    chunks = list(zip(*datas))
+
+    for i, _ in enumerate(chunks):
+        chunks[i] = (chunks[i][0], chunks[i][1])
+
+    properties = {'energies': batch_energies.flatten()}
+
+    # return: [chunk1, chunk2, ...], {"energies", "force", ...} in which chunk1=(species, coordinates)
+    # e.g. chunk1 = [[1807, 21], [1807, 21, 3]], chunk2 = [[193, 50], [193, 50, 3]]
+    # 'energies' = [2000, 1]
+    return chunks, properties
+
+
+def split_to_two_chunks(counts, chunk_threshold):
+
+    counts = counts.cpu()
+    left_mask = torch.triu(torch.ones([counts.shape[0], counts.shape[0]], dtype=torch.uint8))
```

Comment:
Shouldn't mask be bool tensor instead of uint8 tensor?

Reply:
```
RuntimeError: "triu" not implemented for 'Bool'
```

Reply:
OK, this is a bug in PyTorch and should be fixed. Could you please add a comment noting this issue? Something like
```python
# NB (@yueyericardo): In principle this dtype should be `torch.bool`, but unfortunately
# `triu` is not implemented for bool tensor right now. This should be fixed when PyTorch
# add support for it.
```